### DISCUSSION TITLE
Troubleshooting Kathara POC with Python 3.11: Client-Server Communication Issue

### DISCUSSION THREAD

**Discussion by MrStaf** [Kidel/Kathara#242] [Q&A] [answered]
Hello,

I'm trying to make a Kathará POC with a simple Rest API lab. To do this, I'm using custom images of Kathará in which I've installed python 3.11.
Then I use them in the lab, with a client.py and server.py as follows:

```python
# server.py
from fastapi import FastAPI
from random import randint

app = FastAPI()

@app.get("/")
async def root(
    start: int = 0,
    end: int = 100
):
    rand_int = randint(start, end)
    return {"data": rand_int}
```

```python
# client.py
import requests

response = requests.get("http://10.0.0.1:8000/")
print(response.json())
```

The Kathará files are such as :

```bash
# lab.conf
client[0]=A
client[image]="usecase3-client"

server[0]=A
server[image]="usecase3-server"

# server.startup
ifconfig eth0 10.0.0.1 netmask 255.255.255.0 up

# client.startup
ifconfig eth0 10.0.0.2 netmask 255.255.255.0 up
python-work ./client.py

# lab.dep
client: server
```
Here's my problem, when I launch the lab, with the line `CMD ["python-work", "client.py"]` in the **usecase3-client** image, 
it crashes, and the terminal closes before displaying any logs.
When I put `python-work client.py` in the `client.startup`, it also crashes, but differently:

```bash
HTTPConnectionPool(host='10.0.0.1', port=8000): Max retries exceed with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff791ea2510>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

But when I run the same command in the terminal right afterwards, it works and I get this:

```bah
$ python-work client.py
{'data': 39}
``` 

I don't really understand how to do it. Do you have a solution so that I can make it work?
If you need more information, feel free to ask :)
Thanks in advance

**Comment by tcaiazzi** [comment]
Hi @MrStaf, 

I need to play with the lab to understand what is happening. 

Can you please attach your lab and the Dockerfile to build your custom image? 

Anyway, today we released the new docker images based on Debian 12. They come with Python3.11 installed :innocent: 

Thanks, 
Tommaso

**Reply by MrStaf** [reply]
Hi @tcaiazzi,

Here's some good news, I had to do a little hacking to get it all working with python 3.11 but here's the lab: 
[UseCase3.zip](https://github.com/KatharaFramework/Kathara/files/12736032/UseCase3.zip)
If you need any further information, don't hesitate to ask :)

**Reply by tcaiazzi** [reply, marked as answer]
Hi @MrStaf, 

I tried the lab :smiley: 

It seems that your problem is caused by the client that queries the server before this is ready. 
If you add the line `CMD ["python-work", "client.py"]` in the **usecase3-client** image the terminal crashes because `client.py` raises an exception since the server is not ready. 

There are many way to circumvent this issue, a simple one could be to force the client to make the request until the server is ready. 
```python
# client.py
import time

import requests


def get_response():
    try:
        return requests.get("http://10.0.0.1:8000/")
    except Exception:
        time.sleep(0.1)
        return get_response()


if __name__ == '__main__':
    print(get_response().json())

```

I'm attaching a refactored version ([UseCase3-refactored.zip](https://github.com/KatharaFramework/Kathara/files/12737591/UseCase3-refactored.zip)) of your lab that also includes some improvements:
- The server and client images are based on the new `kathara/base` image (already with `Python3.11`) 
- There is no need to copy the `client.py` and `server.py` files in the Dockerfiles. Indeed, if you create a folder with the name of a device, at the lab startup all the files and directories in that folder are copied in the device by Kathará (maintaining the paths).
For example,  in the following image the `client.py` file will be copied in `/app/client.py` in the client, and the `server.py` file will be copied in `/app/server.py` in the server.
![image](https://github.com/KatharaFramework/Kathara/assets/26346836/2e433aa1-8fd2-4b5b-b345-8720f556c388)

- It is better to move the `CMD ["python3", "-m", "uvicorn", "server:app", "--host", "0.0.0.0", "--port", "8000"]` from the server Dockerfile to the startup commands to avoid hiding important details. 

```python
ifconfig eth0 10.0.0.1 netmask 255.255.255.0 up

python3 -m uvicorn server:app --host 0.0.0.0 --port 8000
```

I hope this helps!

**Reply by MrStaf** [reply]
Thanks a lot ! 
Indeed, this is nice upgrade for my lab :)

I think your solution to this issue is very simple and practical, but don't you think there is a more elegant way ? It seems that the timeout delay of the request isn't working as it should.

I'm confused right now, because as I was checking your response, I wanted to remove the warning `Waiting startup commands execution. Press [ENTER] to override...`, so I changed the `server.startup` file from :

```bash
ifconfig eth0 10.0.0.1 netmask 255.255.255.0 up
python3 -m uvicorn server:app --host 0.0.0.0 --port 8000 
```
to

```bash
ifconfig eth0 10.0.0.1 netmask 255.255.255.0 up
nohup python3 -m uvicorn server:app --host 0.0.0.0 --port 8000 & 
```

By using `nohup [...] &` it seems like I unlocked the command, and now it works as I expected :)

![image](https://github.com/KatharaFramework/Kathara/assets/65668723/178ded7d-dbef-47b3-96c2-a5d804cfc374)

**Reply by tcaiazzi** [reply]
There are certainly more elegant ways; mine is just an example. 

The message `Waiting startup commands execution. Press [ENTER] to override...` signals that the device is still executing startup commands. It is needed to give control to users only when the startup commands are finished and the device configured.  

The usage of `nohup python3 -m uvicorn server:app --host 0.0.0.0 --port 8000 & ` is more or less equivalent to press enter on the device shell without waiting the commands. 

I hope this clarify some doubts. In case you need further help, don't hesitate to ask. 

P.S. If you are satisfied, please mark the reply as answer :innocent: 

Thanks!

**Reply by MrStaf** [reply]
Thanks a lot again, it's all clear :)
